### PR TITLE
Add a script for evaluation trained agents on datasets (#19)

### LINKED ISSUE #19: Add script for evaluating an agent on a large dataset
The script should:
* allow loading a pre-trained agent, an image dataset as well as a set of ground truth bounding boxes
* let the agent act on the images from the dataset, logging its predicted bounding boxes
* compare the predicted bounding boxes with the ground truth (maybe using ChainerCV's IoU calculation?)
* output the logged results in a human and machine readable way

### PR DESCRIPTION
**Pull request by bakoe**
Closes #19

### CODE REVIEW

File: evaluate_agent.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    relative_paths = np.loadtxt(imagefile, dtype=str)
+    images_base_path = os.path.dirname(imagefile)
+    absolute_paths = [images_base_path + i.strip('.') for i in relative_paths]
+
+    images = np.loadtxt(imagefile, dtype=str).tolist()
+
+    gt_bboxes = np.load(boxfile)
+
+    env = TextLocEnv(absolute_paths, gt_bboxes, gpu)
```
Comment (J-L-O) [inline evaluate_agent.py]: You could use the create_environment method from the load_agent file instead

**Review by J-L-O — APPROVED** [review]
Looks good and seems to work as well. I haven't completed my test run yet though, since it takes so much time.

File: load_agent.py [hunk 2]
```diff
@@ -50,6 +41,15 @@ def load_agent(env, directory="agent", gpu=0, epsilon=0.3):
     return agent
 
 
+def create_environment(imagefile='image_locations.txt', boxfile='bounding_boxes.npy', gpu=0):
+    locations = np.loadtxt(imagefile, dtype=str).tolist()
+    bboxes = np.load(boxfile)
+
+    env = TextLocEnv(locations, bboxes, gpu)
+
+    return env
```
Comment (milanpro) [inline load_agent.py]: maybe emit variable here?

Comment (milanpro) [reply load_agent.py]: This is not a change of yours it seems. Maybe change it anyways? :)

**Review by milanpro — APPROVED** [review]
Well done!

File: evaluate_agent.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+@click.command()
+@click.option("--gpu", default=-1, help="ID of the GPU to be used. -1 if the CPU should be used instead.")
+@click.option("--imagefile", "-i", default='image_locations.txt', help="Path to the file containing the image locations.", type=click.Path(exists=True))
+@click.option("--boxfile", "-b", default='bounding_boxes.npy', help="Path to the bounding boxes.", type=click.Path(exists=True))
+@click.option("--agentdirectory", "-a", default='./agent', help="Path to the directory containing the agent that should be evaluated.", type=click.Path(exists=True))
+@click.option("--save/--dont-save", default=False, help="Boolean indicating whether the intermediate results of the evaluation should be saved.")
+def evaluate(gpu, imagefile, boxfile, agentdirectory, save):
+    max_steps_per_image = 200
+    max_steps_per_run = 40
```
Comment (FMalerba) [inline evaluate_agent.py]: shouldn't max_steps_per_run = 50 considering that on train_agent.py the max_episode_len was 50?

File: evaluate_agent.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        pred_scores.append(scores)
+
+    if save:
+        np.save('gt_bboxes.npy', gt_bboxes)
+        np.save('gt_labels.npy', gt_labels)
+        np.save('pred_bboxes.npy', pred_bboxes)
+        np.save('pred_labels.npy', pred_labels)
+        np.save('pred_scores.npy', pred_scores)
+
+    eval = eval_detection_voc(pred_bboxes, pred_labels, pred_scores, gt_bboxes, gt_labels)
```
Comment (FMalerba) [inline evaluate_agent.py:78]: Ok, in our case this is not a problem because we have only one bounding box, but if there was more than one how does this precision metric deal with it? I mean Assume that you have 5 gt_bboxes in a given image and assume that the agent finds them all with IoU of 1.0, but it finds them in a different order (say, 2,4,3,0,1). Will this be counted as precision of 100% (correct precision) or will it think that it got them all wrong?

File: evaluate_agent.py (hunk 1 above)
Comment (bakoe) [reply evaluate_agent.py]: Hmm … I wasn't quite sure about that since the `create_environment` method does not convert relative paths to absolute paths. But maybe it's a good time to change the `create_environment` accordingly to be more robust and then use it here in `evaluate_agent.py`. I'll do that.

File: evaluate_agent.py (hunk 3 above)
Comment (bakoe) [reply evaluate_agent.py]: Good point – I'll change that. Thanks for noticing!

File: load_agent.py (hunk 2 above)
Comment (bakoe) [reply load_agent.py]: 👍 I'll change it.

File: evaluate_agent.py (hunk 4 above)
Comment (bakoe) [reply evaluate_agent.py:78]: Actually I'm not too sure about that either. According to the [documentation comments inside `eval_detection_voc`'s source code](https://github.com/chainer/chainercv/blob/master/chainercv/evaluations/eval_detection_voc.py#L11), I'd assume that it should work for multiple bounding boxes per image as well. I haven't tested this, though.

Comment (bakoe) [reply evaluate_agent.py:78]: As discussed: I’m marking this as resolved since you successfully tested it. Thanks!
